go-controller/pkg/clustermanager/egressip_controller.go

```diff
@@ -1723,3 +1711,146 @@ func (eIPC *egressIPClusterController) removePendingOpsAndGetResyncs(egressIPNam
+type jsonPatchOperation struct {
+	Operation string      `json:"op"`
+	Path      string      `json:"path"`
+	Value     interface{} `json:"value,omitempty"`
+}
+
+// patchEgressIP performs a patch operation on an EgressIP.
+// There are two possible patches operations.
+// 1. Mandatory, replace operation of egress IP status field. This allows us to
+// update only the status field, without overwriting any other. This is
+// important because processing egress IPs can take a while (when running on a
+// public cloud and in the worst case), hence we don't want to perform a full
+// object update which risks resetting the EgressIP object's fields to the state
+// they had when we started processing the change.
+// 2. Optional, add operation to its metadata.annotations field.
+func (eIPC *egressIPClusterController) patchEgressIP(name string, patches ...jsonPatchOperation) error {
+	klog.Infof("Patching status on EgressIP %s: %v", name, patches)
+	op, err := json.Marshal(patches)
+	if err != nil {
+		return fmt.Errorf("error serializing patch operation: %+v, err: %v", patches, err)
+	}
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		return eIPC.kube.PatchEgressIP(name, op)
+	})
+}
+
+// generateEgressIPPatches conditionally generates a mark patch if the mark doesn't exist. If it fails to allocate a mark,
+// log an error instead of failing because we do not wish to block primary default network egress IP assignments due to potential
+// mark range exhaustion. Primary default network egress IP currently does not utilize marks to config EgressIP.
+// Generating the status patch is mandatory
+func (eIPC *egressIPClusterController) generateEgressIPPatches(name string, annotations map[string]string,
+	statusItems []egressipv1.EgressIPStatusItem) []jsonPatchOperation {
+	patches := make([]jsonPatchOperation, 0, 1)
+	if !util.IsEgressIPMarkSet(annotations) {
+		if mark, err := eIPC.getOrAllocMark(name); err != nil {
+			klog.Errorf("Failed to get mark for EgressIP %s: %v", name, err)
+		} else {
+			patches = append(patches, generateMarkPatchOp(mark))
+		}
+	}
+	return append(patches, generateStatusPatchOp(statusItems))
+}
+
+func generateMarkPatchOp(mark int) jsonPatchOperation {
+	return jsonPatchOperation{
+		Operation: "add",
+		Path:      "/metadata/annotations",
+		Value:     createAnnotWithMark(mark),
+	}
+}
+
+func createAnnotWithMark(mark int) map[string]string {
+	return map[string]string{util.EgressIPMarkAnnotation: fmt.Sprintf("%d", mark)}
+}
+
+func generateStatusPatchOp(statusItems []egressipv1.EgressIPStatusItem) jsonPatchOperation {
+	return jsonPatchOperation{
+		Operation: "replace",
+		Path:      "/status",
+		Value: egressipv1.EgressIPStatus{
+			Items: statusItems,
+		},
+	}
+}
+
+// syncEgressIPMarkAllocator iterates over all existing EgressIPs. It builds a mark cache of existing marks stored on each
+// EgressIP annotation or allocates and adds a new mark to an EgressIP if it doesn't exist
```

Comment:
when you allocate and add a new mark how do you base the algorithm off the base?

Reply:
I dont think I understand your question but..
Heres how it works.
CM starts up and reads all the EIP objects from the kapi and builds a cache of already allocated marks. It then can assign new marks that do not conflict with old marks.